### PR TITLE
feature/INTERLOK-3463

### PR DESCRIPTION
**Pull request by albinoloverats**
## Motivation

Currently customers are combining several services to make a request and parse the response, whereas the Data Lake consumer and producer are provided to simplify that process, making use of the Microsoft APIs.

## Modification

Add Data Lake consumer/producer. Although I have also added to the Outlook365 tests 

## Result

Instead of having to combine several services to put/get data from Data Lake, there is now a consumer and producer.

## Testing

An Azure account, with data storage capability is necessary to *live test* this, otherwise it's just the mocked unit tests.

### CODE REVIEW

**Comment by quotidian-ennui** [comment]
@albinoloverats  

Looks generally good. 
- I would change your xstream aliases to be azure-XXX just to be explicit about it.
- Please add javadocs etc to all the configurable classes along with the `@config`  javadoc annotation.

Also, I hope your secrets that are in the sources; aren't actually real secrets, cos they're out there now!.

**Comment by quotidian-ennui** [comment]
also guava-30-jre (there's a dependabot PR)

**Comment by albinoloverats** [comment]
> 
> 
> @albinoloverats
> 
> Looks generally good.
> 
>     * I would change your xstream aliases to be azure-XXX just to be explicit about it.
> 
>     * Please add javadocs etc to all the configurable classes along with the `@config`  javadoc annotation.
> 
> 
> Also, I hope your secrets that are in the sources; aren't actually real secrets, cos they're out there now!.

Will do.

And no, they're not actual ID's/secrets.

**Review by aaron-mcgrath-adp — APPROVED** [review]
Agreed with the previous;
"azure-*" as the config element names and javadoc especially around the username so people know what it maps to.

**Comment by quotidian-ennui** [comment]
Also, can you provide an upload service component into Datalake (it can just silently wrap the producer like HttpRequestService does); since this would replicate what Rauf has written.

**Review by aaron-mcgrath-adp — COMMENTED** [review]
On the service:
I reckon if we make the connection an AdaptrisConnection, then we can use shared stuff.
Also probably don't need line 99, that's probably handled in the standalone producers init phase.

I do wonder though, do we even need this service?  Just use a standalone-producer instead?
Having said that it's quite nice to have an explicit service...

**Review by quotidian-ennui — COMMENTED** [review]
> 
> I do wonder though, do we even need this service? Just use a standalone-producer instead?
> Having said that it's quite nice to have an explicit service...

standalone-producer doesn't render nicely in the UI... for most things like this (e.g. making an HTTP request, uploading to S3, etc) it's an activity that happens as part of the workflow, not as a termination of the workflow; which means that most of the time you'll be doing it as a service.

File: interlok-azure-datalake/src/main/java/com/adaptris/interlok/azure/datalake/DataLakeUploadService.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+/**
+ * Upload files to Microsoft Data Lake.
+ *
+ * @config azure-data-lake-upload-service
+ */
+@XStreamAlias("azure-data-lake-upload-service")
+@AdapterComponent
+@ComponentProfile(summary = "Put data into a Azure Data Lake", tag = "service,azure,data lake,data,lake")
+@DisplayOrder(order = { "fileSystem", "path", "filename" })
+public class DataLakeUploadService extends ServiceImp
```
[the remark this answers is not in the document]
Comment (albinoloverats) [reply interlok-azure-datalake/src/main/java/com/adaptris/interlok/azure/datalake/DataLakeUploadService.java]: Done.

File: interlok-azure-datalake/src/main/java/com/adaptris/interlok/azure/datalake/DataLakeUploadService.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  @Getter
+  @Setter
+  @NotBlank
+  @InputFieldHint(expression = true)
+  private String filename;
+  
+  /**
+   * {@inheritDoc}.
+   */
+  @Override
```
[the remark this answers is not in the document]
Comment (albinoloverats) [reply interlok-azure-datalake/src/main/java/com/adaptris/interlok/azure/datalake/DataLakeUploadService.java]: Done - copied from recently updated One Drive services.

File: interlok-azure-datalake/src/main/java/com/adaptris/interlok/azure/datalake/DataLakeUploadService.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+   *
+   * @param adaptrisMessage The message to upload.
+   * @throws ServiceException If there was an issue uploading the file.
+   */
+  @Override
+  public void doService(AdaptrisMessage adaptrisMessage) throws ServiceException
+  {
+    try
+    {
+      DataLakeProducer producer = new DataLakeProducer();
```
[the remark this answers is not in the document]
Comment (albinoloverats) [reply interlok-azure-datalake/src/main/java/com/adaptris/interlok/azure/datalake/DataLakeUploadService.java]: Looks like I may have already made this change.

**Review by quotidian-ennui — APPROVED** [review]

**Pull request merged**